### ISSUE TITLE
virtual bool AstCallNode::initRegisters(codeGen&): Assertion `callee' failed. Aborted (core dumped)

### ISSUE THREAD

**Issue by 1Husin-Alhaj** [dyninst/dyninst#442]
I tested the Instrumenting a function example (from Dyninst Programmer’s Guide 2016 ) bu I always get this error:
patch: /home/husien/Downloads/dyninst-master/dyninstAPI/src/ast.C:2069: virtual bool AstCallNode::initRegisters(codeGen&): Assertion `callee' failed.
Aborted (core dumped)
I am trying to re-write the binary file on disk.
I have compiled the patch.C file using -ldyninstAPI Ccorrectly without any error.
![op](https://user-images.githubusercontent.com/17450624/37776925-067cafde-2e04-11e8-9d2d-539c662f281c.jpg)

My operating system is Linux Ubunut 14. and I am using Dyninst-master.

**Comment by mxz297** [comment]
I will take a look at your problem today. I am sorry that I was not able to respond promptly.

**Comment by 1Husin-Alhaj** [comment]
I am very grateful to you
If you want more details I will explain the problem here in detail.
Just a simple example that I am trying to rewrite the binary file by loading a function from a shared libray, but always the same problem occurs. I tested several code from the web and from the Dyninst Programmer’s Guide 2016 but ther is no response. I'm just interested in static binary instrumentation. All the programs that I have tested, are based on BPatch_binaryedit. More than 4 times I removed and re-installed Dynist but the problem still appear.  

Thank you very much

**Comment by mxz297** [comment]
I am not able to reproduce your problem, assuming you are still using the examples we discussed last time. I am able to correct run "patch" to rewrite the "hello" program and new generated binary shows correct output. Can you package all your source code and binaries, such patch.cpp, Makefile, hello.cpp, hello and send it to me?

**Comment by 1Husin-Alhaj** [comment]
Yes, sure.
I explained all the steps that I did

[Dyninst-test-case.zip](https://github.com/dyninst/dyninst/files/1839517/Dyninst-test-case.zip)

**Comment by mxz297** [comment]
I noticed that your "patch" file is 32-bit. Are you running on a 32-bit machine or a 64-bit machine? Do you compile Dyninst in 64-bit or 32-bit?

**Comment by 1Husin-Alhaj** [comment]
I am working on Linux Ubuntu 32-bit. Therefore I have compiled Dyninst in 32-bit.

**Comment by mxz297** [comment]
I see. My machine is 64 bit. I will compile Dyninst in 32bit mode and see whether I can reproduce your problem.

**Comment by 1Husin-Alhaj** [comment]
Ok :)

**Comment by 1Husin-Alhaj** [comment]
Please Can you send to me your Cmake command lines that you will type in order to compile and build Dyninst? 
If we can't find a solution, I will remove it and re-compile Dyninst. However, with cmake command line, we should pass some configuring options. What are the most important, not popular, options?
If this is not important, I will wait you. If the code run correctly with Dyninst 32-bit, please tell me what is the version and all characteristics of the machine you use. Maybe, I will force to change the operating system and replace it with another version.
Thank you very much.

**Comment by mxz297** [comment]
My machine is a 64-bit machine running Ubuntu 16.04.4. I highly recommend that you install ccmake because the GUI interfaces provide clear meaning of used CMAKE variables and show their default values. If you use ubuntu, ccmake can be easily installed by "sudo apt-get install cmake-curses-gui"

At the same time, I found it was not trivial to compile 32-bit Dyninst on a 64-bit machine. I will continue to try or find a 32-bit environment. But if you have access to a 64-bit machine, I am sure it would be better for you to switch to a 64-bit environment.

**Comment by pstach** [comment]
I encountered what looks like this same issue.  As it turns out the environment variable DYNINSTAPI_RT_LIB needed to be pointed at the correct runtime.

**Comment by 1Husin-Alhaj** [comment]
Finally, I detected the mistake. Unfortunately, it is a trifle mistake. I have changed my machine from 32-bit to 64-bit. Dyninst doesn't care either machine 32 or 64 bit. The error was in setting  the environment variable DYNINSTAPI_RT_LIB. 
error command line: export DYNINSTAPI_RT_LIB=/usr/local/lib/libdyninstAPI.so
correct command line:export DYNINSTAPI_RT_LIB=/usr/local/lib/libdyninstAPI_RT.so

we should point to libdyninstAPI_RT.so instead of libdyninstAPI.so.
Thanks for help. 
this issue is closed.
